hicommonwealth/commonwealth
# Cosmos Proposal Submitted Notifications

## Link to Issue
Closes: #4864

## Description of Changes
**Main Changes**
_These changes can be found in `packages/commonwealth/server/workers/cosmosGovNotifications/`_

Added a function which fetches proposals for every Cosmos chain that has at least 1 subscription and creates notifications from the proposals. There are 2 possible branches of execution:
1. A prior notification is found for the given chain. In this case, we extract the proposal id from the notification and try to fetch every proposal from that id onwards. We generate a notification for all of those proposals as long as they are not older than 2 hours.
2. No prior notification is found for the given chain. In this case, we simply fetch the latest proposal from the chain and generate a notification for it if it was created less than 2 hours ago.

Generating notifications for proposals that have been created in the last 2 hours guarantees that we will never miss emitting a notification unless the server is down for longer than 2 hours (extremely unlikely).

**Other Changes**
- Removed block_number as a requirement for chain-event notifications
- Updated the database reset function used in testing to use `bulkCreate` rather than `Promise.all(...create)`
- Updated Cosmos devnet test filenames to match repo pattern

## Test Plan
- Run `yarn emit-notification -c kyve -w [your-wallet-address] -m`
- Ensure you can see the notification in your notifications dropdown when logged in with the associated wallet address.

## Deployment Plan

1. Deploy to CW.
3. Create a Heroku scheduler function to run `yarn send-cosmos-notifs` every 10 minutes.

## Other Considerations

- #4943 should be a fast follow so that these notifications become actually useful.
